### DISCUSSION TITLE
Where to put code for validation checks across multiple lines?

### DISCUSSION THREAD

**Discussion by SteveBronder** [stan-dev/stanc3#934] [Q&A]
I'd like to check that a transformed parameter is ever used before it's assigned, but I'm not sure where to put that kind of check?

In #930 I'm working on an optimization that would let us not fill objects with NA values as long as they are fully assigned to before use. @rok-cesnovar found a very good example [link](https://github.com/stan-dev/stanc3/pull/930#issuecomment-886052750) where my current schema fails because the program _never_ assigns to the variable. I can fix that, but in general I think it would be nice to fail programs that have variables defined in transformed parameters that are never assigned to as that's most likely a program error. But looking at the semantic checks, particularly `semantic_check_ostatements_in_block` it's not clear to me how to perform multi-line validation statements. What would be the best way to include checks that cover multiple lines like what I'm discussing here? @rok-cesnovar also pointed out that this error is actually caught in pedantic model. Maybe we can import some schemes from pedantic mode to the actual validation checking?

**Comment by rok-cesnovar** [comment]
I think this is already separated via the `--warn-uninitialized` flag and maybe we should just do a better job of promoting that flag to the users? Unless we think this was tested enough and could be used as default?

Warning as opposed to stoping with and error seems safer as I guess there could be cases of false positives?

**Reply by SteveBronder** [reply]
I think it would be great to have that on by default, then at least users can also report places to us where we think something is uninitialized when it actually is initialized. I especially like it on by default if it's only a warning, we could even include a message like, "If you believe this variable is initialized and this warning is a false positive please file an issue at the stanc3 repo (link)"

**Comment by nhuurre** [comment]
I agree --warn-uninitialized would be good to have on by default but last time we discussed this Ryan said false alarms should not be considered bugs.
https://github.com/stan-dev/stanc3/issues/912#issuecomment-856119386
